<a href="https://colab.research.google.com/github/manoharpavuluri/dl-sonar-chirp-boucing/blob/main/sonar.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data',header=None)
df

0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      51      52      53      54      55      56      57  \
0    0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1    0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2    0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3    0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4    0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
..      ...  ...     ...     ...     ...     ...     ...     ...     ...   
203  0.2684  ...  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065  0.0115   
204  0.2154  ...  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034  0.0032   
205  0.2529  ...  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140  0.0138   
206  0.2354  ...  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034  0.0079   
207  0.2354  ...  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040  0.0036   

         58      59  60  
0    0.0090  0.0032   R  
1    0.0052  0.0044   R  
2    0.0095  0.0078   R  
3    0.0040  0.0117   R  
4    0.0107  0.0094   R  
..      ...     ...  ..  
203  0.0193  0.0157   M  
204  0.0062  0.0067   M  
205  0.0077  0.0031   M  
206  0.0036  0.0048   M  
207  0.0061  0.0115   M  

[208 rows x 61 columns]

In [4]:
df.describe(include='all')

0           1           2           3           4   \
count   208.000000  208.000000  208.000000  208.000000  208.000000   
unique         NaN         NaN         NaN         NaN         NaN   
top            NaN         NaN         NaN         NaN         NaN   
freq           NaN         NaN         NaN         NaN         NaN   
mean      0.029164    0.038437    0.043832    0.053892    0.075202   
std       0.022991    0.032960    0.038428    0.046528    0.055552   
min       0.001500    0.000600    0.001500    0.005800    0.006700   
25%       0.013350    0.016450    0.018950    0.024375    0.038050   
50%       0.022800    0.030800    0.034300    0.044050    0.062500   
75%       0.035550    0.047950    0.057950    0.064500    0.100275   
max       0.137100    0.233900    0.305900    0.426400    0.401000   

                5           6           7           8           9   ...  \
count   208.000000  208.000000  208.000000  208.000000  208.000000  ...   
unique         NaN         NaN         NaN         NaN         NaN  ...   
top            NaN         NaN         NaN         NaN         NaN  ...   
freq           NaN         NaN         NaN         NaN         NaN  ...   
mean      0.104570    0.121747    0.134799    0.178003    0.208259  ...   
std       0.059105    0.061788    0.085152    0.118387    0.134416  ...   
min       0.010200    0.003300    0.005500    0.007500    0.011300  ...   
25%       0.067025    0.080900    0.080425    0.097025    0.111275  ...   
50%       0.092150    0.106950    0.112100    0.152250    0.182400  ...   
75%       0.134125    0.154000    0.169600    0.233425    0.268700  ...   
max       0.382300    0.372900    0.459000    0.682800    0.710600  ...   

                51          52          53          54          55  \
count   208.000000  208.000000  208.000000  208.000000  208.000000   
unique         NaN         NaN         NaN         NaN         NaN   
top            NaN         NaN         NaN         NaN         NaN   
freq           NaN         NaN         NaN         NaN         NaN   
mean      0.013420    0.010709    0.010941    0.009290    0.008222   
std       0.009634    0.007060    0.007301    0.007088    0.005736   
min       0.000800    0.000500    0.001000    0.000600    0.000400   
25%       0.007275    0.005075    0.005375    0.004150    0.004400   
50%       0.011400    0.009550    0.009300    0.007500    0.006850   
75%       0.016725    0.014900    0.014500    0.012100    0.010575   
max       0.070900    0.039000    0.035200    0.044700    0.039400   

                56          57          58          59   60  
count   208.000000  208.000000  208.000000  208.000000  208  
unique         NaN         NaN         NaN         NaN    2  
top            NaN         NaN         NaN         NaN    M  
freq           NaN         NaN         NaN         NaN  111  
mean      0.007820    0.007949    0.007941    0.006507  NaN  
std       0.005785    0.006470    0.006181    0.005031  NaN  
min       0.000300    0.000300    0.000100    0.000600  NaN  
25%       0.003700    0.003600    0.003675    0.003100  NaN  
50%       0.005950    0.005800    0.006400    0.005300  NaN  
75%       0.010425    0.010350    0.010325    0.008525  NaN  
max       0.035500    0.044000    0.036400    0.043900  NaN  

[11 rows x 61 columns]

In [5]:
#will return name of columns that contains null rows - looks like none. good!

df.columns[df.isnull().any()].tolist()

[]

In [6]:
# lets see distinct values in the target column ie column 60

df[60].value_counts()



M    111
R     97
Name: 60, dtype: int64

* M stands for mineral

* R stands for rock

In [7]:
from re import A
# lets create x and y

X = df.drop(60,axis=1)
y = df[60]

y.head()


0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [8]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [9]:
# one hot encoding on column 60 to convert R and M to integer
# and since it has only 2 values,  drop 1 and it will automatically assign the other 0

y = pd.get_dummies(y,drop_first=True)
y.sample(10)




R
94   1
97   0
73   1
84   1
83   1
203  0
88   1
4    1
37   1
204  0

In [10]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [11]:
df[60].value_counts()

M    111
R     97
Name: 60, dtype: int64

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25,random_state=1)

In [16]:
X_train.shape, X_test.shape

((156, 60), (52, 60))

In [17]:
import tensorflow as tf
from tensorflow import keras

In [19]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 7s 6ms/step - loss: 0.7318 - accuracy: 0.4615
Epoch 2/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6879 - accuracy: 0.5513
Epoch 3/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6830 - accuracy: 0.6603
Epoch 4/100
20/20 [==============================] - 0s 6ms/step - loss: 0.6755 - accuracy: 0.6795
Epoch 5/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6684 - accuracy: 0.5897
Epoch 6/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6576 - accuracy: 0.6410
Epoch 7/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6424 - accuracy: 0.6538
Epoch 8/100
20/20 [==============================] - 0s 7ms/step - loss: 0.6237 - accuracy: 0.6987
Epoch 9/100
20/20 [==============================] - 0s 6ms/step - loss: 0.6088 - accuracy: 0.6474
Epoch 10/100
20/20 [==============================] - 0s 4ms/step - loss: 0.5973 - accuracy: 0.6987
Epoch 11/

In [21]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 9ms/step - loss: 0.8926 - accuracy: 0.7500


[0.8925988078117371, 0.75]

In [26]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print('predict : ', y_pred[:10])
print('test : ', y_test[:10].transpose())

2/2 [==============================] - 0s 6ms/step
[0.00217101 0.9998185  0.5682718  0.01594166 1.         0.99996877
 0.96084046 1.         0.00672134 1.        ]
predict :  [0. 1. 1. 0. 1. 1. 1. 1. 0. 1.]
test :     186  155  165  200  58   34   151  18   202  62 
R    0    0    0    0    1    1    0    1    0    1


In [23]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [27]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.73      0.81      0.77        27
           1       0.77      0.68      0.72        25

    accuracy                           0.75        52
   macro avg       0.75      0.75      0.75        52
weighted avg       0.75      0.75      0.75        52





```
## Model with Dropout Layer
```



In [28]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 6s 8ms/step - loss: 0.7038 - accuracy: 0.5513
Epoch 2/100
20/20 [==============================] - 0s 8ms/step - loss: 0.7235 - accuracy: 0.4936
Epoch 3/100
20/20 [==============================] - 0s 7ms/step - loss: 0.7335 - accuracy: 0.5321
Epoch 4/100
20/20 [==============================] - 0s 8ms/step - loss: 0.7140 - accuracy: 0.4808
Epoch 5/100
20/20 [==============================] - 0s 7ms/step - loss: 0.7167 - accuracy: 0.5128
Epoch 6/100
20/20 [==============================] - 0s 7ms/step - loss: 0.7061 - accuracy: 0.5128
Epoch 7/100
20/20 [==============================] - 0s 7ms/step - loss: 0.7012 - accuracy: 0.5064
Epoch 8/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6988 - accuracy: 0.4808
Epoch 9/100
20/20 [==============================] - 0s 6ms/step - loss: 0.6858 - accuracy: 0.5385
Epoch 10/100
20/20 [==============================] - 0s 5ms/step - loss: 0.6979 - accuracy: 0.5385
Epoch 11/

In [29]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 16ms/step - loss: 0.4227 - accuracy: 0.7500


[0.4226545989513397, 0.75]

In [30]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 [==============================] - 0s 11ms/step
[2.3646250e-04 7.5331920e-01 6.5945065e-01 2.2162694e-02 9.9861932e-01
 8.7768859e-01 6.8658966e-01 9.9880159e-01 3.3701200e-02 9.9919957e-01]
[0. 1. 1. 0. 1. 1. 1. 1. 0. 1.]


In [31]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.72      0.85      0.78        27
           1       0.80      0.64      0.71        25

    accuracy                           0.75        52
   macro avg       0.76      0.75      0.75        52
weighted avg       0.76      0.75      0.75        52



You can see that by using dropout layer test accuracy increased from 0.77 to 0.81